dashboard/test/ui/runner.rb

```diff
@@ -282,22 +287,58 @@ def output_synopsis(output_text)
     end
   end
 
+  def how_many_reruns?(test_run_string)
+    if $options.auto_retry
+      return 1
+    elsif $options.magic_retry
+      # ask saucelabs how flaky the test is
+      flakiness = TestFlakiness.test_flakiness[test_run_string]
+      if !flakiness
+        $lock.synchronize { puts "No flakiness data for #{test_run_string}".green }
+        return 0
+      elsif flakiness == 0.0
+        $lock.synchronize { puts "#{test_run_string} is not flaky".green }
+        return 0
+      else
+        flakiness_message = "#{test_run_string} is #{flakiness} flaky. "
+        max_reruns = (1 / Math.log(flakiness, 0.05)).ceil - 1 # reruns = runs - 1
+        confidence = (1.0 - flakiness ** (max_reruns + 1)).round(3)
+        flakiness_message +=  "we should rerun #{max_reruns} times for #{confidence} confidence"
+
+        if max_reruns < 2
+          $lock.synchronize { puts flakiness_message.green }
+        elsif max_reruns < 3
+          $lock.synchronize { puts flakiness_message.yellow }
+        else
+          $lock.synchronize { puts flakiness_message.red }
+        end
+        return max_reruns
+      end
+    else # no retry options
+      return 0
+    end
+  end
+
+  max_reruns = how_many_reruns?(test_run_string)
+
   # if autorertrying, output a rerun file so on retry we only run failed tests
   rerun_filename = test_run_string + ".rerun"
-  first_time_arguments = $options.auto_retry ? " --format rerun --out #{rerun_filename}" : ""
+  first_time_arguments = (max_reruns > 0) ? " --format rerun --out #{rerun_filename}" : ""
 
   FileUtils.rm rerun_filename, force: true
 
   succeeded, output_stdout, output_stderr, test_duration = run_tests(arguments + first_time_arguments)
```

Comment:
Would it make sense to fold `arguments` into  `first_time_arguments` now that they're never used separately (I think?)